Google.PowerShell.IntegrationTests/BigQuery/BigQuery.BqTabledata.Tests.ps1

```diff
@@ -93,4 +93,113 @@ Describe "Set-BqSchema" {
+        $filename_broken_csv = "$folder\classics_broken.csv"
+        $filename_broken_json = "$folder\classics_broken.json"
+        # This file has a missing field and some of the AVRO formatting has been deleted.
+        $filename_broken_avro = "$folder\classics_broken.avro"
+    }
+
+    BeforeEach {
+        $r = Get-Random
+        $table = New-BqTable -Dataset $test_Set "table_$r"
+        $table = New-BqSchema -Name "Title" -Type "STRING" |
+                 New-BqSchema -Name "Author" -Type "STRING" |
+                 New-BqSchema -Name "Year" -Type "INTEGER" |
+                 Set-BqSchema $table
+    }
+
+    It "should properly consume a well formed CSV file" {
+        $table | Add-BqTabledata $filename_csv CSV -SkipLeadingRows 1
+        $table = Get-BqTable $table
+        $table.NumRows | Should Be 10
+    }
+
+    It "should properly consume a well formed JSON file" {
+        $table | Add-BqTabledata $filename_json JSON
+        $table = Get-BqTable $table
+        $table.NumRows | Should Be 10
+    }
+
+    It "should properly consume a well formed AVRO file" {
+        $table | Add-BqTabledata $filename_avro AVRO
+        $table = Get-BqTable $table
+        $table.NumRows | Should Be 10
+    }
+
+    It "should properly reject an invalid CSV file" {
+        { $table | Add-BqTabledata $filename_broken_csv CSV -SkipLeadingRows 1 -ErrorAction Stop } | Should Throw "contained errors"
+    }
+
+    It "should properly reject an invalid JSON file" {
+        { $table | Add-BqTabledata $filename_broken_json JSON -ErrorAction Stop } | Should Throw "contained errors"
+    }
+
+    It "should properly reject an invalid AVRO file" {
+        { $table | Add-BqTabledata $filename_broken_avro AVRO -ErrorAction Stop } | Should Throw "contained errors"
+    }
+
+    #TODO(ahandley): Test all CSV options
+
+    It "should handle less than perfect CSV files" {
+        $table | Add-BqTabledata $filename_broken_csv CSV -SkipLeadingRows 1 -MaxBadRecords 4 `
+                -AllowUnknownFields -ErrorAction SilentlyContinue
+        $table = Get-BqTable $table
+        $table.NumRows | Should Be 7
+    }
+
+    It "should handle write disposition WriteAppend" {
+        $table | Add-BqTabledata $filename_csv CSV -SkipLeadingRows 1 
+        $table = Get-BqTable $table
+        $table.NumRows | Should Be 10
+        $table | Add-BqTabledata $filename_csv CSV -SkipLeadingRows 1 -WriteMode WriteAppend
+        $table = Get-BqTable $table
+        $table.NumRows | Should Be 20
+    }
+
+    It "should handle write disposition WriteIfEmpty" {
+        $table | Add-BqTabledata $filename_broken_csv CSV -SkipLeadingRows 1 -MaxBadRecords 4 `
+                -AllowUnknownFields -ErrorAction SilentlyContinue
+        $table = Get-BqTable $table
+        $table.NumRows | Should Be 7
+        { $table | Add-BqTabledata $filename_csv CSV -SkipLeadingRows 1 `
+            -WriteMode WriteIfEmpty -ErrorAction Stop } | Should Throw "404"
```

Comment:
We should test scenario where user use -WriteIfEmpty and no error is thrown too. Same for -WriteTruncate and -WriteAppend.